src/legacy/implode.rs

```diff
@@ -0,0 +1,277 @@
+                is.advance(1 + used)?;
+            } else {
+                sym = lsb(bits, 8) as u16;
+                is.advance(1 + 8)?;
+            }
+            debug_assert!(sym <= u8::MAX as u16);
+            dst.push_back(sym as u8);
+            continue;
+        }
+        // Backref.
+        debug_assert!(lsb(bits, 1) == 0x0);
+        let mut used_tot = 1;
+        bits >>= 1;
+
+        // Read the low dist bits.
+        let mut dist;
+        if large_wnd {
+            dist = lsb(bits, 7) as usize;
+            bits >>= 7;
+            used_tot += 7;
+        } else {
+            dist = lsb(bits, 6) as usize;
+            bits >>= 6;
+            used_tot += 6;
+        }
+
+        // Read the Huffman-encoded high dist bits.
+        let mut used = 0;
+        let sym = dist_decoder.huffman_decode(!bits as u16, &mut used)?;
+        used_tot += used;
+        bits >>= used;
+        dist |= (sym as usize) << if large_wnd { 7 } else { 6 };
+        dist += 1;
+
+        // Read the Huffman-encoded len.
+        let sym = len_decoder.huffman_decode(!bits as u16, &mut used)?;
+        used_tot += used;
+        bits >>= used;
+        let mut len = (sym + min_len) as usize;
+
+        if sym == 63 {
+            // Read an extra len byte.
+            len += lsb(bits, 8) as usize;
+            used_tot += 8;
+            //  bits >>= 8;
+        }
+
+        debug_assert!((used_tot as usize) <= ISTREAM_MIN_BITS);
+        is.advance(used_tot)?;
+        //  let len = len.min(uncomp_len - dst.len());
+
+        if len <= uncomp_len - dst.len() && dist <= dst.len() {
+            // Enough room and no implicit zeros; chunked copy.
+            lz77_output_backref(dst, dist, len);
+        } else {
+            // Copy, handling overlap and implicit zeros.
+            for _i in 0..len {
+                if dist > dst.len() {
+                    dst.push_back(0);
+                    continue;
+                }
+                dst.push_back(dst[dst.len() - dist]);
+            }
+        }
+    }
+
+    *src_used = is.bytes_read();
+    Ok(())
+}
+
+#[derive(Debug)]
+pub struct ImplodeDecoder<R> {
+    compressed_reader: R,
+    uncompressed_size: u64,
+    stream_read: bool,
+    large_wnd: bool,
+    lit_tree: bool,
+    stream: VecDeque<u8>,
+}
+
+impl<R: Read> ImplodeDecoder<R> {
+    pub fn new(inner: R, uncompressed_size: u64, flags: u16) -> Self {
+        let large_wnd = (flags & 2) != 0;
+        let lit_tree = (flags & 4) != 0;
+        ImplodeDecoder {
+            compressed_reader: inner,
+            uncompressed_size,
+            stream_read: false,
+            large_wnd,
+            lit_tree,
+            stream: VecDeque::new(),
+        }
+    }
+
+    pub fn finish(mut self) -> Result<VecDeque<u8>> {
+        copy(&mut self.compressed_reader, &mut self.stream)?;
+        Ok(self.stream)
+    }
+}
+
+impl<R: Read> Read for ImplodeDecoder<R> {
+    fn read(&mut self, buf: &mut [u8]) -> Result<usize> {
```

Comment:
This implementation looks good as long as `buf.len() >= self.stream.len()`, but if not then we need one that doesn't truncate.

Reply:
I don't think that decompressing the whole file to memory is something I like.

Reply:
That's fine, you can use a smaller buffer -- or no buffer, if that turns out to perform better. I just thought you'd prefer to use the all-at-once implementation where it was applicable, given that it's probably the simplest one.

Reply:
The all-at-once approach was just the simplest one to port over. It's good enough for my use case.

I wasn't sure if there is interest in taking these algorithms to the zip-rs crate. I didn't port the compression bits (I don't see any sense in that and code needs to be maintained as well) and I didn't know if it's in the scope of zip-rs at all.

Reply:
reverted the commit. Didn't make sense to me TBH.